### PR TITLE
Add tooling to spin up local environments

### PR DESCRIPTION
**Pull request by simu**
## Summary

Add shell script and make targets to spin up a kind-based local environment. The setup script prompts the user to setup a dev realm and admin user in an existing Keycloak instance. The PR doesn't provide tooling to spin up a local Keycloak instance at the moment.

## Checklist

- [x] PR contains a single logical change (to build a better changelog).
- [x] Update the documentation.
- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog.

### CODE REVIEW

**Comment by ccremer** [comment]
Just another source of inspiration: In K8up we also did a KIND setup, but based on `make`: https://github.com/k8up-io/k8up/tree/master/e2e

**Comment by tobru** [comment]
Vcluster uses k3s, here we're using kind: We should at least document that so that there is awareness that it isn't the same Kubernetes distro.

**Comment by simu** [comment]
> Vcluster uses k3s, here we're using kind: We should at least document that so that there is awareness that it isn't the same Kubernetes distro.

I don't really care what we use, but [APUB-155 (internal)](https://ticket.vshn.net/browse/APUB-155) explicitly asks for kind

File: Makefile [hunk 1]
```diff
@@ -56,6 +58,14 @@ build.docker: $(BIN_FILENAME) ## Build the docker image
 clean: ## Cleans up the generated resources
 	rm -rf dist/ cover.out $(BIN_FILENAME) || true
 
+.PHONY: local-env
+local-env-setup: ## Setup local kind-based dev environment
+	$(localenv_make) setup
+
+.PHONY: local-env-clean-setup
+local-env-clean-setup: ## Setup local kind-based dev environment
```
Comment (glrf) [inline Makefile:66]: The comment is wrong and IMO the tag misleading. I would just call it `local-env-clean`

File: README.md [hunk 2]
```diff
@@ -17,14 +17,24 @@ This can be done with make:
...
+
+## Local development environment
+
+You can setup a kind-based local environment with
+
+```bash
+local-env/setup-kind.sh
+```
+
+See the [local-env/README.md](./local-env/README.md) for more details on the local environment setup.
```
Comment (glrf) [inline README.md:40]: This isn't correct? We need to call the script through `make`?

File: local-env/README.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+## Installation
+
+The `setup.sh` script will guide you through the setup.
+There are some steps that you have to perform manually on a Keycloak instance, which the script prompts you for.
+The script defaults to VSHN's APPUiO Dev Keycloak instance, but you can provide an URL pointing to a different instance during the install process.
+
+```bash
+./setup.sh
+```
```
Comment (glrf) [inline local-env/README.md:24]: The script isn't called `setup.sh` but `kind-setup.sh`. Also it takes arguments. I would suggest just referencing the Makefile instead.

File: local-env/setup-kind.sh [hunk 4]
```diff
@@ -0,0 +1,84 @@
+#!/bin/bash
+# vim:sts=2:ts=2:et:sw=2:tw=0
+
+readonly script_dir=$(dirname "$0")
+readonly kind_cmd="${1}"
+readonly kind_cluster="${2}"
+readonly kind_node_version="${3}"
+readonly kind_kubeconfig="${4}"
```
Comment (glrf) [inline local-env/setup-kind.sh:8]: We should either fail fast or provide a default if these are not provided.

File: local-env/setup-kind.sh [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+while [ x"$identifier" == x"" ]; do
+  read -r -p "Provide an identifier for your local-dev Keycloak realm: " identifier
+done
+
+realm_name="local-dev-${identifier}"
+sed -e "s/REPLACEME/${realm_name}/g" "${script_dir}/templates/realm.json.tpl" > "${script_dir}/realm.json"
+
+step "Navigate to ${keycloak_url} and create a new realm by importing the '$(realpath "${script_dir}/realm.json")' file."
+
+step "Create a user in the new realm, grant it realm role 'admin', and ensure 'Email Verified' is set to 'On'."
```
Comment (glrf) [inline local-env/setup-kind.sh:47]: Could we not provide a default admin user for a realm?

File: local-env/setup-kind.sh [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+step "Note: In the next step, a browser window will open where you have to sign in to Keycloak with the user you've created in the previous step".
```
Comment (glrf) [inline local-env/setup-kind.sh:49]: This was IMO a bit confusing. That way you get a note with "hey a browser window will open". But it won't, you first need to hit enter. I would prefer a (visually more noticeable) note that is not blocking for user input.

**Review by glrf — CHANGES_REQUESTED** [review]
Works pretty good, I was able to setup a test environment fairly quickly.

I only really care about the README change to reference the make command. The others are just nitpicks.

File: local-env/setup-kind.sh (hunk 5 above)
Comment (simu) [reply local-env/setup-kind.sh:47]: I didn't find a way to export users when exporting a realm, but if you know a way I'm all ears

Comment (bastjan) [inline local-env/setup-kind.sh:47]: A note that an email has to be filled in would be great here. I created a user "admin" with no email, but with 'Email Verified' set to On. That is not enough.

Comment (glrf) [reply local-env/setup-kind.sh:47]: No, that was just my naive assumption that that should be possible :sweat_smile:

File: local-env/setup-kind.sh [hunk 7]
```diff
@@ -0,0 +1,84 @@
+#!/bin/bash
```
Comment (bastjan) [inline local-env/setup-kind.sh:1]: Maybe exit on error would reduce some debug sessions?

**Review by bastjan — COMMENTED** [review]
It worked for me with `make -C local-env setup` after some trial and error.

The admin user has to have an email. Just toggling to email verified is not enough.

File: local-env/setup-kind.sh (hunk 7 above)
Comment (bastjan) [reply local-env/setup-kind.sh:1]: The script still works for me with `set -euo pipefail`

File: local-env/setup-kind.sh [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+kubectl config set-credentials oidc-user \
+  --exec-api-version=client.authentication.k8s.io/v1beta1 \
+  --exec-command=kubectl \
+  --exec-arg=oidc-login \
+  --exec-arg=get-token \
+  --exec-arg=--oidc-issuer-url="${keycloak_url}/auth/realms/${realm_name}" \
+  --exec-arg=--oidc-client-id=local-dev \
+  --exec-arg=--oidc-extra-scope="email offline_access profile openid"
+kubectl config set-context --current --user=oidc-user
+kubectl apply -k "${script_dir}/../config/crd/apiextensions.k8s.io/v1"
```
Comment (bastjan) [inline local-env/setup-kind.sh:82]: This does not work for me. I think we should make the admin role the default in dev realms.

File: local-env/setup-kind.sh (hunk 6 above)
Comment (simu) [reply local-env/setup-kind.sh]: Changed the note to be bold (should work in all respectable terminal emulators) and be fenced by lines of equal signs

File: local-env/setup-kind.sh [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+check_command() {
+  if ! command -v "${1}" >/dev/null 2>&1; then
+    step "Install ${2}. Follow the instructions at ${3}"
+  fi
+}
+
+check_command "kubectl" "kubectl" "https://kubernetes.io/docs/tasks/tools/#kubectl"
+check_command "kubectl-oidc_login" "kubectl oidc-login plugin" "Follow the instructions at https://github.com/int128/kubelogin#setup"
```
Comment (ccremer) [inline local-env/setup-kind.sh:30]:
```suggestion
check_command "kubectl-oidc_login" "kubectl oidc-login plugin" "https://github.com/int128/kubelogin#setup"
```

File: local-env/setup-kind.sh (hunk 8 above)
Comment (bastjan) [reply local-env/setup-kind.sh:88]: Nevermind i missread the instructions while setting up the cluster 🙈

**Review by bastjan — APPROVED** [review]
LGTM. The instructions and guided steps now make it quite easy to setup the dev environment! 🧙‍♂️

File: local-env/setup-kind.sh [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+  --exec-command=kubectl \
+  --exec-arg=oidc-login \
+  --exec-arg=get-token \
+  --exec-arg=--oidc-issuer-url="${keycloak_url}/auth/realms/${realm_name}" \
+  --exec-arg=--oidc-client-id=local-dev \
+  --exec-arg=--oidc-extra-scope="email offline_access profile openid"
+kubectl config set-context --current --user=oidc-user
+kubectl apply -k "${script_dir}/../config/crd/apiextensions.k8s.io/v1"
+
+echo "Setup finished. Set environment variable KUBECONFIG to '$(realpath "${kind_kubeconfig}")' to interact with the local dev cluster"
```
Comment (ccremer) [inline local-env/setup-kind.sh:90]: This should be better highlighted, hard to miss in the first-time output. Also, provide a copy-pastable command
```suggestion
echo =======
echo "Setup finished. To interact with the local dev cluster, set the KUBECONFIG environment variable"
echo "\"export \$KUBECONFIG=$(realpath ${kind_kubeconfig})\""
echo =======
```

**Review by ccremer — APPROVED** [review]
It worked well for me. Just small suggestions

**Review by glrf — APPROVED** [review]

**Pull request merged**